### PR TITLE
Feedback on caching constructed Nullable<T> types

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Changes made in response to feedback after prior PRs were already merged.
Relates to https://github.com/dotnet/roslyn/pull/67766

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1596]: >typeArgument.IsEnumType() [](http://example.com/codeflow?start=20&length=25)

Why is it a good idea to cache for enum types? This is not a closed set of types.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1596]: Because it's the largest source of constructing these types as all the work is done around enum operators and lifted operators.  It's not a closed set.  But they're also a tiny minority of all types, so caching does not produce much here, but saves a large amount of overall allocations.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1596]: For context, enums are by far the heaviest source of construct allocations during typing: 

image

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1596]: > Because it's the largest source of constructing these types as all the work is done around enum operators

Doesn't this depend mostly on the user code?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1596]: Effectively, an external consumer, regardless of what is in the source code, can grow this cache unbounded.

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi What is the status of this PR?

**Comment by CyrusNajmabadi** [comment]
Effectively waiting on you or other compiler feedback.  Not sure what to do with your enum feedback.  I'm personally unconcerned about the pathological case of someone continuing to throw enums at this.  I'm more concerned about the practical case and the expense around operators and enums.  I'd be fine seeing how the existing code in the compiler does in practice today and refining if we run into any issues with that.

**Comment by AlekseyTs** [comment]
> Not sure what to do with your enum feedback.

I think we should limit the caching. Either by specific known set of types, or my the size of the cache. Or revert the cache completely.

**Comment by CyrusNajmabadi** [comment]
> I think we should limit the caching. Either by specific known set of types, or by the size of the cache. 

I don't like either of these.  Limiting to a set of known types defeats the purpose.  LImiting the size of hte cache also seems speculative and unnecessary absent real world cases of concern.  #wontfix.

> Or revert the cache completely.

I'm ok with the cache as is, and would prefer to only change it based on real world cases that arise.

**Comment by CyrusNajmabadi** [comment]
Moving to draft while we consider a path forward.

**Comment by AlekseyTs** [comment]
@CyrusNajmabadi Is this PR ready for review?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Is this PR ready for review?

Yup, this is ready.  Thanks for the suggestions.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1584]: Nit: I think it would be worth including the type that failed in the assertion for debugging purposes.

**Review by 333fred — COMMENTED** [review]
LGTM (commit 6), one minor suggestion.

**Review by 333fred — APPROVED** [review]

**Pull request merged**
